### PR TITLE
Extensible statistics

### PR DESCRIPTION
**Pull request by asilvas**
Tried to keep scope down but the changes are pretty considerable and come with tradeoffs. The big downside of leveraging a consistent stats interface across all layers is that distributed training now requires a bit more effort to process the http results from the remote models. I believe the tradeoffs are well worth it though, and the updated docs attempt to explain these new patterns that should enable some pretty incredible and robust stats in the future (that I also plan to contribute).

## Statistics

```mermaid
graph TD
  OpA(Op A) --> statsA{{"stats A"}};
  OpB(Op B) --> statsA;
  statsA --> LoggerA{{"LoggerConsole A"}};
  LoggerA --> Stdout(("Stdout"));
  OpC(Op C) --> statsA;
  OpD(Op D) --> statsA;
  statsA --> LoggerB("LoggerCustom B");
  LoggerB --> Disk(("Disk"));
```

Basic usage of gathering statistics is as simple adding
a collector using the default `StatsLoggerConsole`.

```
import { stats, StatsLoggerConsole, rand, matmul } from '@shumai/shumai'

stats.enabled = true // all ops following will capture stats

// perform ops...

stats.enabled = false // all ops following will no longer capture stats
```

While the above examples may suffice for simple use cases, if you're
looking to capture stats across multiple threads, processes, and/or hosts,
`StatsLoggerHttp` has you covered.

```mermaid
graph TD
  subgraph Host C
    Processor("LoggerHttp Processor")
    style Processor stroke:#222,stroke-width:4px,stroke-dasharray:5 5
  end
  subgraph Host A
    OpA(Op A) --> statsA{{"stats A"}};
    OpB(Op B) --> statsA;
    statsA --> LoggerA{{"LoggerHttp A"}};
    LoggerA --> Processor;
  end
  subgraph Host B
    OpC(Op C) --> statsB{{"stats B"}};
    OpD(Op D) --> statsB;
    statsB --> LoggerB{{"LoggerHttp B"}};
    LoggerB --> Processor;
  end
```

```
import { StatsLoggerHttp } from '@shumai/shumai'

stats.logger = new StatsLoggerHttp({ url: 'http://localhost:4242' })
```

For more custom needs you can supply your own logger:

```
import { StatsLogger, StatsLoggerData } from '@shumai/shumai'

class CustomLogger implements StatsLogger {
  async process(data: StatsLoggerData): Promise<void> {
    const summary = data.collector.getSummary()
    console.log('Collector stats:', summary)
  }
}

stats.logger = new CustomLogger()
```

By default stack tracing is disabled as it adds 50%+ overhead, but can be enabled via `stats.collectStacks = true`.

### Scoped Statistics

If you wish to isolate stats profiling you can do this as well:

```
import { collectStats } from '@shumai/shumai'

const scopedStats = collectStats(() => {
  // perform ops...
}/*, StatsCollectorOptions | StatsLogger */)
console.log(scopedStats.getSummary())
```

### CODE REVIEW

**Comment by bwasti** [comment]
my main question is, would it be possible to maintain "requires_stats" as a concept while adding this functionality?  I typically only profile periodically and would like fine grained control over how that's done.

I suspect the "collectors" can still be registered in global scope and any requires_stats tensor will lookup and feed into them.

**Comment by asilvas** [comment]
> my main question is, would it be possible to maintain "requires_stats" as a concept while adding this functionality? I typically only profile periodically and would like fine grained control over how that's done.
> 
> I suspect the "collectors" can still be registered in global scope and any requires_stats tensor will lookup and feed into them.

What is the motive to tie stats directly with tensors? The existing implementation has a fair amount of overhead and complexity as it required collections and merges/copies. My objectives were to streamline stats across both simple and complex use cases, and to provide an extensible solution that could integrate externally (think `shumai-profiler`). I want to be sure all the intended use cases were covered, just not sure how `t.require_stats` is doing anything more useful than this proposal.

**Comment by bwasti** [comment]
> What is the motive to tie stats directly with tensors?

It's just a convenient way to "signal" because it's functional (with respect to user exposed state) and it scales to many machines.

> The existing implementation has a fair amount of overhead and complexity as it required collections and merges/copies.

Agreed

> not sure how `t.require_stats` is doing anything more useful than this proposal.

To clarify, I don't think it conflicts too much with your proposal.  It's purely an API/semantics addition.  I'd like to keep it accessible to all programmers and `requires_grad` is super well understood.  Here's how I'd expect it to work:

Disabled:
```mermaid
 graph LR
   Tensor(Tensor, requires_stats=false);
   subgraph Host A
     Tensor -.-> OpA;
     OpA(Op A) -.-> OpB;
     OpB(Op B);
   end
  OpB -.-> OpC;
   subgraph Host B
     OpC(Op C) -.-> OpD;
     OpD(Op D);
   end
 ```

Enabled:
```mermaid
 graph LR
   Tensor(Tensor, requires_stats=true);
   subgraph Host C
     Processor("LoggerHttp Processor")
     style Processor stroke:#222,stroke-width:4px,stroke-dasharray:5 5
   end
   subgraph Host A
     Tensor -.-> OpA;
     OpA(Op A) -.-> OpB;
     OpB(Op B);
     OpA --> CollectorA{{"Collector A"}};
     OpB --> CollectorA{{"Collector A"}};
     CollectorA --> LoggerA{{"LoggerHttp A"}} --> Processor;
   end
  OpB -.-> OpC;
   subgraph Host B
     OpC(Op C) -.-> OpD;
     OpD(Op D);
     OpC --> CollectorB{{"Collector B"}};
     OpD --> CollectorB{{"Collector B"}};
     CollectorB --> LoggerB{{"LoggerHttp B"}} --> Processor;    
   end
 ```

**Comment by bwasti** [comment]
the missing feature at the moment is about collection over the network

e.g.:
```javascript
let model = sm.network.remote_model(url)

for (let i of sm.util.viter(1000000)) {
  const [x, y] = get_pair()

  // periodically sample from the runner when we start using it
  const profile = i % 1000 === 0 && i < 10000
  x.requires_stats = profile
  
  const y_hat = await model(x)
  await loss(y_hat, y).backward()

  if (profile) {
    const perf = await (await fetch(`${url}/statistics`)).json()
    // some check to see if we are overloading this machine
    if (perf.bytesUsed > 1e9) {
      model = connect_new_model()
    }
  }
}
```

I guess scope could satisfy this, but you'll need to tell the remote model runner to collect statistics for that run.

**Comment by asilvas** [comment]
I see, so if I understand correctly if the network code had the ability to collect the remote stats this would satisfy your concerns? As of right now this is already possible, it's just less convient since it requires each of those remote models gather their own stats and use a remote logger (ala `StatsLoggerHttp`), but if the network code natively supported transferring those stats that would unblock this PR?

So in your example the code might look something like:

```
let model = sm.network.remote_model(url)

for (let i of sm.util.viter(1000000)) {
  const [x, y] = get_pair()

  // periodically sample from the runner when we start using it
  const profile = i % 1000 === 0 && i < 10000
  
  const y_hat = await model(x, { collectStats: profile })
  await loss(y_hat, y).backward()

  if (profile) {
    const perf = await (await fetch(`${url}/statistics`)).json()
    // some check to see if we are overloading this machine
    if (perf.bytesUsed > 1e9) {
      model = connect_new_model()
    }
  }
}
```

File: examples/distributed/model.ts [hunk 1]
```diff
@@ -4,9 +4,10 @@ const model_a = sm.network.remote_model('0.0.0.0:3001')
 const model_b = sm.network.remote_model('0.0.0.0:3002')
 
 const model = async (t: sm.Tensor) => {
-  t = await model_a(t)
-  t = await model_b(t)
-  return t
+  let res: sm.network.TFetchResult
+  res = await model_a(t)
+  res = await model_b(res.tensor)
+  return res.tensor
```
Comment (bwasti) [inline examples/distributed/model.ts:10]: today remote models have the nice property of

```javascript
import model from 'my_script'

const val = await model(t)
```

working exactly the same way as

```javascript
const model = sm.network.remote_model('0.0.0.0:3002')

const val = await model(t)
```

This is very much intentional for debugging ease.  I think we should preserve it.  To add stats, though, we can do this:

```javascript
const val = await model(t)
const stats = await model.stats
```

which will return `null` in the local model case.

Comment (asilvas) [reply examples/distributed/model.ts]: Now that we support the new `Stats` interface on a tensor I was able to restore the old interface `t = await model(t)`

**Comment by bwasti** [comment]
thanks for all the effort you put into this! appreciate you sticking with it despite the back and forth.

I think it looks ready to land :)

**Pull request merged**